src/sdk/encrypt.ts

```diff
@@ -1,94 +1,200 @@
+        )
+          throw new Error('The value must be a boolean, a number or a bigint.');
+        if (
+          (typeof value !== 'bigint' || typeof value !== 'number') &&
+          Number(value) > 1
+        )
+          throw new Error('The value must be 1 or 0.');
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[1]);
+        return this;
+      },
+      add4(value: number | bigint) {
+        checkEncryptedValue(value, 4);
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[4]);
+        return this;
+      },
+      add8(value: number | bigint) {
+        checkEncryptedValue(value, 8);
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[8]);
+        return this;
+      },
+      add16(value: number | bigint) {
+        checkEncryptedValue(value, 16);
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[16]);
+        return this;
+      },
+      add32(value: number | bigint) {
+        checkEncryptedValue(value, 32);
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[32]);
+        return this;
+      },
+      add64(value: number | bigint) {
+        checkEncryptedValue(value, 64);
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[64]);
+        return this;
+      },
+      add128(value: number | bigint) {
+        checkEncryptedValue(value, 128);
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[128]);
+        return this;
+      },
+      addAddress(value: string) {
+        if (!isAddress(value)) {
+          throw new Error('The value must be a valid address.');
+        }
+        values.push(BigInt(value));
+        bits.push(ENCRYPTION_TYPES[160]);
+        return this;
+      },
+      getValues() {
+        return values;
+      },
+      getBits() {
+        return bits;
+      },
+      resetValues() {
+        values.length = 0;
+        bits.length = 0;
+        return this;
+      },
+      encrypt() {
+        const encrypted = CompactFheUint160List.encrypt_with_compact_public_key(
+          values,
+          publicKey,
+        );
+        // const encrypted = ProvenCompactFheUint160List.encrypt_with_compact_public_key(
+        //   values,
+        //   publicZkParams,
+        //   publicKey,
+        //   ZkComputeLoad.Proof,
+        // );
+        const data = encrypted.serialize();
+        const inputs = bits.map((v, i) => {
+          const dataWithIndex = new Uint8Array(data.length + 1);
+          dataWithIndex.set(data, 0);
+          dataWithIndex.set([i], data.length);
+          const hash = createKeccakHash('keccak256')
+            .update(Buffer.from(dataWithIndex))
+            .digest();
+          const dataInput = new Uint8Array(32);
+          dataInput.set(hash, 0);
+          dataInput.set([i, bits[v], 0], 29);
+          return createKeccakHash('keccak256')
+            .update(toHexString(dataInput))
+            .digest('hex');
```

Comment:
Why do we hash `dataInput` again? AFAIU, the handle is supposed to be `dataInput` itself in the form of:

```
keccak256(data) || i || type || version
```

It could be that I am misreading the code, please correct me if I am wrong.